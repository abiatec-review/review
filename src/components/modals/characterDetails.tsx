import TextRow from '@components/textRow';
import useOrientation from '@hooks/useOrientation';
import Character from '@models/character';
import React from 'react';
import {Image, Platform, StyleSheet, View} from 'react-native';

import Modal from '.';

interface Props {
  character: Character;
  isShown: boolean;
  toggle: () => void;
}

function CharacterDetailsModal(props: Props) {
  const {character, isShown, toggle} = props;

  const {isPortrait} = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapeStyles;

  return (
    <Modal isShown={isShown} toggle={toggle} style={styles.modal}>
      <View style={styles.infoBlock}>
        <Image source={{uri: character.image}} style={styles.image} />
        <View style={styles.textBlock}>
          <TextRow field="Status" data={character.status} />
          <TextRow field="Gender" data={character.gender} />
          <TextRow field="Origin" data={character.origin?.name} />
          <TextRow field="Created" data={character.created} />
        </View>
      </View>
    </Modal>
  );
}

export default CharacterDetailsModal;

const baseStyles = StyleSheet.create({
  image: {
    aspectRatio: 1,
    borderRadius: 20,
  },
  textBlock: {
    flex: 1,
    justifyContent: 'space-evenly',
  },
});

const portraitStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: Platform.select({ios: 0.65, android: 0.85}),
  },
  infoBlock: {flex: 1},
  image: {
    ...baseStyles.image,
    width: '100%',
  },
});

const landscapeStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: 0.62,
  },
  infoBlock: {
    marginBottom: 10,
    alignItems: 'center',
    flexDirection: 'row',
  },
  image: {
    ...baseStyles.image,
    flex: Platform.select({ios: 0.45, android: 0.5}),
    marginRight: 15,
  },
  textBlock: {
    ...baseStyles.textBlock,
    height: "80%"
  }
});
