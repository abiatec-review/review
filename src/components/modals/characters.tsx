import useOrientation from '@hooks/useOrientation';
import Character from '@models/character';
import React from 'react';
import * as RN from 'react-native';

import Modal from '.';

interface Props {
  title: string;
  isShown: boolean;
  toggle: () => void;
  characters: Array<Character>;
}

function CharactersModal(props: Props) {
  const {characters, isShown, toggle, title} = props;

  const {isPortrait} = useOrientation();

  const getHeight = () => {
    if (characters.length > 10) {
      return {height: isPortrait ? 270 : 180};
    } else {
      const multiplier = characters.length % 10;
      if (isPortrait) return {height: multiplier > 3 ? 270 : multiplier * 90};
      else return {height: multiplier > 3 ? 180 : multiplier * 90};
    }
  };

  return (
    <Modal isShown={isShown} toggle={toggle} style={{flex: 0}}>
      <RN.Text style={styles.title}>{title}</RN.Text>
      <RN.ScrollView style={getHeight()}>
        {characters.map(({id, name, image}) => (
          <RN.View key={id} style={styles.character}>
            <RN.Image source={{uri: image}} style={styles.image} />
            <RN.Text style={styles.text}>{name}</RN.Text>
          </RN.View>
        ))}
      </RN.ScrollView>
    </Modal>
  );
}

export default CharactersModal;

const styles = RN.StyleSheet.create({
  title: {
    fontSize: 20,
    paddingBottom: 10,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  character: {
    marginBottom: 10,
    alignItems: 'center',
    flexDirection: 'row',
  },
  image: {
    width: 80,
    height: 80,
    marginRight: 10,
    borderRadius: 50,
  },
  text: {
    flex: 1,
    fontSize: 18,
    fontWeight: '600',
  },
});
