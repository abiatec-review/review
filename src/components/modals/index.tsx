import useOrientation from '@hooks/useOrientation';
import React from 'react';
import * as RN from 'react-native';

interface Props {
  isShown: boolean;
  toggle: () => void;
  style?: RN.FlexStyle;
}

function Modal(props: React.PropsWithChildren<Props>) {
  const {isShown, toggle, children, style} = props;

  const {isPortrait} = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapelStyles;

  return (
    <RN.Modal
      transparent
      animationType="fade"
      visible={isShown}
      onRequestClose={toggle}
      supportedOrientations={['portrait', 'landscape']}>
      <RN.View style={styles.container}>
        <RN.View style={[styles.content, style]}>
          {children}
          <RN.Pressable style={styles.closeButton} onPress={toggle}>
            <RN.Text style={styles.closeText}>Close</RN.Text>
          </RN.Pressable>
        </RN.View>
      </RN.View>
    </RN.Modal>
  );
}

export default Modal;

const baseStyles = RN.StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  content: {
    padding: 20,
    borderRadius: 20,
    backgroundColor: 'white',
  },
  closeButton: {
    width: '100%',
    padding: 10,
    borderRadius: 20,
    backgroundColor: '#2196F3',
  },
  closeText: {
    fontSize: 18,
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

const portraitStyles = RN.StyleSheet.create({
  ...baseStyles,
  content: {
    ...baseStyles.content,
    width: '91%',
  },
});

const landscapelStyles = RN.StyleSheet.create({
  ...baseStyles,
  content: {
    ...baseStyles.content,
    width: '70%',
  },
});
