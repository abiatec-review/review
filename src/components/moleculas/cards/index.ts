export { CharacterCard } from "./character";
export { LocationCard } from "./location";
export { EpisodeCard } from "./episode";
