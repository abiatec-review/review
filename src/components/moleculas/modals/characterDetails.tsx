import React from "react";

import { Image, Platform, StyleSheet, Text, View } from "react-native";

import { TextRow } from "@components/atoms";
import { Modal } from "@components/atoms";
import { useOrientation } from "@hooks";
import { Character } from "@redux/models/entities";
import { FontSize, Indent, Radius } from "@utils";

interface Props {
  character: Character;
  isShown: boolean;
  toggle: () => void;
}

export function CharacterDetailsModal(props: Props) {
  const { character, isShown, toggle } = props;

  const { isPortrait } = useOrientation();
  const styles = isPortrait ? portraitStyles : landscapeStyles;

  return (
    <Modal isShown={isShown} toggle={toggle} style={styles.modal}>
      <Text style={styles.name}>{character.name}</Text>
      <View style={styles.infoBlock}>
        <Image source={{ uri: character.image }} style={styles.image} />
        <View style={styles.textBlock}>
          <TextRow field="Status" data={character.status} />
          <TextRow field="Gender" data={character.gender} />
          <TextRow field="Origin" data={character.origin?.name} />
          <TextRow field="Created" data={character.created} />
        </View>
      </View>
    </Modal>
  );
}

const baseStyles = StyleSheet.create({
  name: {
    fontWeight: "bold",
    textAlign: "center",
    fontSize: FontSize.MEDIUM
  },
  image: {
    aspectRatio: 1,
    borderRadius: Radius.DEFAULT
  },
  infoBlock: {
    flex: 1
  },
  textBlock: {
    flex: 0.9,
    justifyContent: "space-evenly"
  }
});

const portraitStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: Platform.select({ ios: 0.65, android: 0.85 })
  },
  name: {
    ...baseStyles.name,
    marginBottom: Indent.DEFAULT
  },
  image: {
    ...baseStyles.image,
    width: "100%"
  }
});

const landscapeStyles = StyleSheet.create({
  ...baseStyles,
  modal: {
    flex: 0.6
  },
  infoBlock: {
    ...baseStyles.infoBlock,
    flexDirection: "row",
    alignItems: "center"
  },
  image: {
    ...baseStyles.image,
    marginRight: Indent.HUGE,
    flex: 0.5
  },
  textBlock: {
    ...baseStyles.textBlock,
    flex: 1,
    height: "80%"
  }
});
