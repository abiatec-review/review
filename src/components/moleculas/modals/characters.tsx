import { useOrientation } from "@hooks";
import { Character } from "@models/entities";
import { FontSize, Indent, Radius } from "@utils";
import React from "react";
import * as RN from "react-native";

import Modal from "./modal";

interface Props {
  title: string;
  isShown: boolean;
  toggle: () => void;
  characters: Array<Character>;
}

export function CharactersModal(props: Props) {
  const { characters, isShown, toggle, title } = props;

  const { isPortrait } = useOrientation();

  const getHeight = () => {
    if (characters.length > 10) return { height: isPortrait ? 270 : 180 };

    const multiplier = characters.length % 10;
    if (isPortrait) return { height: multiplier > 3 ? 270 : multiplier * 90 };
    else return { height: multiplier > 3 ? 180 : multiplier * 90 };
  };

  return (
    <Modal isShown={isShown} toggle={toggle} style={{ flex: 0 }}>
      <RN.Text style={styles.title}>{title}</RN.Text>
      <RN.ScrollView style={getHeight()}>
        {characters.map(({ id, name, image }) => (
          <RN.View key={id} style={styles.character}>
            <RN.Image source={{ uri: image }} style={styles.image} />
            <RN.Text style={styles.text}>{name}</RN.Text>
          </RN.View>
        ))}
      </RN.ScrollView>
    </Modal>
  );
}

const styles = RN.StyleSheet.create({
  title: {
    fontWeight: "bold",
    textAlign: "center",
    fontSize: FontSize.MEDIUM,
    paddingBottom: Indent.DEFAULT
  },
  character: {
    flexDirection: "row",
    alignItems: "center",
    marginBottom: Indent.DEFAULT
  },
  image: {
    width: 80,
    height: 80,
    borderRadius: Radius.HUGE,
    marginRight: Indent.DEFAULT
  },
  text: {
    flex: 1,
    fontWeight: "600",
    fontSize: FontSize.MEDIUM
  }
});
