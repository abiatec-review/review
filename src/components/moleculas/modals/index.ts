export { CharactersModal } from "./characters";
export { CharacterDetailsModal } from "./characterDetails";
