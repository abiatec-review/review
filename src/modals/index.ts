export { CharacterFullInfo } from './CharacterFullInfo';
export { ImagePickerModal } from './ImagePickerModal';
