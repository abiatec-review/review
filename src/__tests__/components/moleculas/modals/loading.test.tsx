import React from "react";

import { create } from "react-test-renderer";

import { LoadingModal } from "@components/moleculas/modals";

jest.mock("@react-native-firebase/auth", () => {});
jest.mock("@react-native-firebase/storage", () => {});
jest.mock("react-native-image-picker", () => {});
jest.mock("react-native-vector-icons/MaterialIcons", () => {});

test("LoadingModal render", () => {
  const tree = create(<LoadingModal />).toJSON();
  expect(tree).toMatchSnapshot();
});
