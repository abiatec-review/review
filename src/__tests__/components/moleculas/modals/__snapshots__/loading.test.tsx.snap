// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingModal render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.5)",
      "bottom": 0,
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#2196F3"
    size="large"
    style={
      Object {
        "margin": 20,
      }
    }
  />
</View>
`;
